src/core/leader_schedule.zig

```diff
@@ -9,205 +9,318 @@ const Pubkey = sig.core.Pubkey;
+                bank,
+            );
+        }
+        return entry.value_ptr.slot_leaders[slot_index];
     }
 
-    const Record = struct { slot: Slot, key: Pubkey };
+    pub fn getSlotLeaderMaybeComputeRpc(self: *LeaderScheduleCache, slot: Slot, rpc_client: *RpcClient) !sig.core.Pubkey {
+        const epoch, const slot_index = self.epoch_schedule.getEpochAndSlotIndex(slot);
+        if (self.getSlotLeader(epoch, slot_index)) |leader| return leader;
 
-    var leaders_index: usize = 0;
-    var leaders = try allocator.alloc(Record, num_leaders);
-    defer allocator.free(leaders);
+        const leader_schedules, var leader_schedule_lg = self.leader_schedules_rw.writeWithLock();
+        defer leader_schedule_lg.unlock();
 
-    var rpc_leader_iter = rpc_leader_schedule.iterator();
-    while (rpc_leader_iter.next()) |entry| {
-        const key = try Pubkey.fromString(entry.key_ptr.*);
-        for (entry.value_ptr.*) |slot| {
-            leaders[leaders_index] = .{ .slot = slot, .key = key };
-            leaders_index += 1;
+        const entry = try leader_schedules.getOrPut(epoch);
+        if (!entry.found_existing) {
+            if (leader_schedules.count() == MAX_CACHED_LEADER_SCHEDULES) {
+                _ = leader_schedules.swapRemove(std.mem.min(Epoch, leader_schedules.keys()));
+            }
+            entry.value_ptr.* = try LeaderSchedule.fromRpc(
+                self.allocator,
+                slot,
+                rpc_client,
+            );
         }
+        return entry.value_ptr.slot_leaders[slot_index];
     }
 
-    std.mem.sortUnstable(Record, leaders, {}, struct {
-        fn gt(_: void, lhs: Record, rhs: Record) bool {
-            return switch (std.math.order(lhs.slot, rhs.slot)) {
-                .gt => false,
-                else => true,
-            };
-        }
-    }.gt);
+    fn getSlotLeader(self: *LeaderScheduleCache, epoch: Epoch, slot_index: usize) ?Pubkey {
+        const leader_schedules, var leader_schedule_lg = self.leader_schedules_rw.readWithLock();
+        defer leader_schedule_lg.unlock();
 
-    var leader_pubkeys = try allocator.alloc(Pubkey, leaders.len);
-    for (leaders, 0..) |record, i| {
-        leader_pubkeys[i] = record.key;
+        if (leader_schedules.getEntry(epoch)) |entry| {
+            return entry.value_ptr.slot_leaders[slot_index];
+        } else {
+            return null;
+        }
     }
 
-    return .{
-        .allocator = allocator,
-        .slot_leaders = leader_pubkeys,
-        .start_slot = start_slot,
-    };
-}
-
-pub const StakedNode = struct { id: Pubkey, stake: u64 };
+    /// This method exists purely to allow using leader schedule from cli, manually inserting it into the
+    /// cache with its associated epoch.
+    pub fn insertLeaderSchedule(self: *LeaderScheduleCache, epoch: Epoch, leader_schedule: LeaderSchedule) !void {
+        const leader_schedules, var leader_schedule_lg = self.leader_schedules_rw.writeWithLock();
```

Comment:
should we have a `if (leader_schedules.count() == MAX_CACHED_LEADER_SCHEDULES) {` check in this method?

Reply:
265ed12